sound/soc/sof/intel/hda-dai.c

```diff
@@ -29,63 +29,86 @@ struct hda_pipe_params {
 	unsigned int link_bps;
 };
 
-/*
- * Unlike GP dma, there is a set of stream registers in hda controller
- * to control the link dma channels. Each register controls one link
- * dma channel and the relation is fixed. To make sure FW uses correct
- * link dma channels, host allocates stream registers and sends the
- * corresponding link dma channels to FW to allocate link dma channel
- *
- * FIXME: this API is abused in the sense that tx_num and rx_num are
- * passed as arguments, not returned. We need to find a better way to
- * retrieve the stream tag allocated for the link DMA
- */
-static int hda_link_dma_get_channels(struct snd_soc_dai *dai,
-				     unsigned int *tx_num,
-				     unsigned int *tx_slot,
-				     unsigned int *rx_num,
-				     unsigned int *rx_slot)
+static bool hda_check_fes(struct snd_soc_pcm_runtime *rtd,
+			  int dir, int stream_tag)
 {
-	struct hdac_bus *bus;
-	struct hdac_ext_stream *stream;
-	struct snd_pcm_substream substream;
-	struct snd_sof_dev *sdev =
-		snd_soc_component_get_drvdata(dai->component);
-
-	bus = sof_to_bus(sdev);
-
-	memset(&substream, 0, sizeof(substream));
-	if (*tx_num == 1) {
-		substream.stream = SNDRV_PCM_STREAM_PLAYBACK;
-		stream = snd_hdac_ext_stream_assign(bus, &substream,
-						    HDAC_EXT_STREAM_TYPE_LINK);
-		if (!stream) {
-			dev_err(bus->dev, "error: failed to find a free hda ext stream for playback");
-			return -EBUSY;
-		}
-
-		snd_soc_dai_set_dma_data(dai, &substream, stream);
-		*tx_slot = hdac_stream(stream)->stream_tag - 1;
-
-		dev_dbg(bus->dev, "link dma channel %d for playback", *tx_slot);
+	struct snd_pcm_substream *fe_substream;
+	struct hdac_stream *fe_hstream;
+	struct snd_soc_dpcm *dpcm;
+
+	/*
+	 * check if BE is connected to the FE with
+	 * the same stream tag.
```

Comment:
and this is needed for....???

Reply:
@plbossart this is needed because of our limitation to surround the DMA format setting with couple and decouple. This happens during the FE hw_params.

So the sequence is:

PCM open -> assign host DMA ch -> link hw_params -> assign link DMA ch -> host hw_params -> couple link and host DMA ch -> set DMA format -> decouple link and host DMA ch

Amd here's the flow: 
1. So when a PCM is opened, the first free host DMA channel is assigned to it. 
2. When we get to link hw_params, we need to find a free link DMA channel but we need to make sure the host DMA channel corresponding to the same link DMA channel is connected to the link. Otherwise, the coupling that happens during host hw_params will break the usecase.

Reply:
I was asking for a short explanation added to this comment...